raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java

```diff
@@ -495,6 +496,31 @@ public void testHandleBeginQuorumEpochAfterUserInitiatedResign() throws Exceptio
             context.listener.currentLeaderAndEpoch());
     }
 
+    @Test
+    public void testBeginQuorumHeartbeat() throws Exception {
+        int localId = 0;
+        int remoteId1 = 1;
+        int remoteId2 = 2;
+        Set<Integer> voters = Utils.mkSet(localId, remoteId1, remoteId2);
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(localId, voters).build();
+
+        context.becomeLeader();
+        assertEquals(OptionalInt.of(localId), context.currentLeader());
+
+        // begin epoch requests should be sent out every beginQuorumEpochTimeoutMs
+        context.time.sleep(context.beginQuorumEpochTimeoutMs);
+        context.client.poll();
+        context.expectBeginEpoch(context.currentEpoch());
+
+        context.client.poll();
+        assertThrows(AssertionFailedError.class, () -> context.expectBeginEpoch(context.currentEpoch()));
```

Comment:
What is this testing?

Reply:
This was originally preceded by a time.sleep(5ms) to check we won't send begin epoch requests before the timeout  expires - but then I realized this could be flaky because we don't know for certain how much time is left on the timer. I think I've thought of a more useful way to check we don't send more requests than expected